---
title: "AW Fisehr tutorial"
author: "Zhiguang Huo"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
    number_sections: true
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = F}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

options(stringsAsFactors = F)

```


# Introduction

A tutorial to guide through the usage for the AWFisher R package.
A real data example of the mltiple tissue mouse metabolism data is used.
The following parts are included:

- How to prepare the input for AWFisher.
- Transcriptomic meta analysis.
- Differential expression pattern detection.

# How to install the package

The package is available on GitHub page ([https://github.com/Caleb-Huo/AWFisher](https://github.com/Caleb-Huo/AWFisher))

To install the package, 

```{r, eval=F}
library(devtools)
install_github("Caleb-Huo/AWFisher")
```

# How to cite the package

The paper is accepted by the annuals of applied statistics.

- Huo, Zhiguang, Shaowu Tang, Yongseok Park, and George Tseng. "P-value evaluation, variability index and biomarker categorization for adaptively weighted Fisher's meta-analysis method in omics applications." arXiv preprint arXiv:1708.05084 (2018). (Submitted)

The pre-print can be found on ArXiv ([https://arxiv.org/pdf/1708.05084.pdf](https://arxiv.org/pdf/1708.05084.pdf))


# How to prepare the input for AWFisher

Note that you may need internet connection to read in the data.

## Include packages
```{r}
# include necessary packages
library(AWFisher) # Include the BayesMP package
library(limma) # Will perform differential expression analysis
```

## Read in data
```{r}
# read in the data

data_brown <- read.csv("https://bayesmp.github.io/data/mouseMetabolism/data_brown.csv", row.names = 1)
data_heart <- read.csv("https://bayesmp.github.io/data/mouseMetabolism/data_heart.csv", row.names = 1)
data_liver <- read.csv("https://bayesmp.github.io/data/mouseMetabolism/data_liver.csv", row.names = 1)

# Verify gene names match across three tissues
all(rownames(data_brown) == rownames(data_heart))
all(rownames(data_brown) == rownames(data_liver))

# Combime these three studies as list
dataExp <- list(brown=data_brown, heart=data_heart, liver=data_liver)

# Check the dimension of the three studies
sapply(dataExp, dim)

# Check the head of the three studies
sapply(dataExp, head)

# perform differential expression analysis for each of these three tissues.

# Create an empty matrix to store Z value. 
# Each row represents a gene and each column represent a study/tissue. 
# Note that Z value matrix is the input for BayesMP method. 
# Z value can be calculated by inverse CDF of the p-value from differential expression analysis. 
# A positive Z value indicates that the gene is upregulated in that study/tissue.

pmatrix <- matrix(0,nrow=nrow(dataExp[[1]]),ncol=length(dataExp))
rownames(pmatrix) <- rownames(dataExp[[1]])
colnames(pmatrix) <- names(dataExp)
```


## Perform differential expression analysis in each study
```{r}
for(s in 1:length(dataExp)){
	adata <- dataExp[[s]]
	ControlLabel = grep('wt',colnames(adata))
	caseLabel = grep('LCAD',colnames(adata))
	label <- rep(NA, ncol(adata))
	label[ControlLabel] = 0
	label[caseLabel] = 1
	
	design = model.matrix(~label)	# design matrix
	fit <- lmFit(adata,design)		# fit limma model
	fit <- eBayes(fit)		
	
	pmatrix[,s] <- fit$p.value[,2]
}

head(pmatrix)
```


# Transcriptomic meta analysis.



```{r}
# WD <- '~/Desktop/'
# setwd(WD) # You can set the working directory here. Some MCMC results will be saved here.

res = AWFisher.pvalue(pmatrix)
qvalue <- p.adjust(res$pvalue, "BH")
sum(qvalue < 0.05) ## differentially expressed genes with FDR 5%
head(res$weights) ## show the AW weight
```

# Differential expression pattern detection.

## prepare data and calculate dissimilarity matrix

```{r}
studies <- NULL
for(s in 1:length(dataExp)){
	adata <- dataExp[[s]]
	ControlLabel = grep('wt',colnames(adata))
	caseLabel = grep('LCAD',colnames(adata))
	label <- rep(NA, ncol(adata))
	label[ControlLabel] = 0
	label[caseLabel] = 1
	
  studies[[s]] <- list(data=adata, label=label)
}



result <- biomarkerCategorization(studies,function_limma,B=100,DEindex=NULL,seed = 15213)
sum(result$DEindex) ## print out DE index at FDR 5%
head(result$varibility) ## print out the head of variability index
print(result$dissimilarity[1:4,1:4]) ## print out the dissimilarity matrix
```

## Apply tight clustering algorithm

```{r}
tightClustResult <- tight.clust(result$dissimilarity, target=4, k.min=15, random.seed=15213)
clusterMembership <- tightClustResult$cluster

```


## Visualization

### Visualize the heatmap of the first metaPattern module
```{r, fig.show='hold'}

for(s in 1:length(dataExp)){
	adata <- dataExp[[s]]
	aname <- names(dataExp)[s]
	bdata <- adata[qvalue<0.05, ][tightClustResult$cluster == 1 ,]
	cdata <- as.matrix(bdata)
	ddata <- t(scale(t(cdata))) # standardize the data such that for each gene, the mean is 0 and sd is 1.

	ColSideColors <- rep("black", ncol(adata))
	ColSideColors[grep('LCAD',colnames(adata))] <- "red"
	
	B <- 16
  redGreenColor <- rgb(c(rep(0, B), (0:B)/B), c((B:0)/16, rep(0, B)), rep(0, 2*B+1))
	heatmap(ddata,Rowv=NA,ColSideColors=ColSideColors,col= redGreenColor ,scale='none',Colv=NA, main=aname)
}
```

